primitives/src/circuit/merkle_tree/rescue_merkle_tree.rs

```diff
@@ -41,7 +41,19 @@ pub struct StandardLeafVar {
     pub elem: Variable,
 }
 
-impl<F> MerkleTreeGadget<RescueMerkleTree<F>> for PlonkCircuit<F>
+impl<F: RescueParameter> DigestAlgorithmGadget<RescueMerkleTree<F>> for PlonkCircuit<F> {
+    fn digest(&mut self, data: &[Variable]) -> Result<Variable, CircuitError> {
```

Comment:
I'm actually little unsure.
Ideally, shouldn't `DigestAlgorithmGadget` only implemented for `MerkleTreeGadget`? We don't need these `digest_leaf` gadget exposing to the general `PlonkCircuit` IMO.

this may solve our `dyn DigestAlgorithmGadget` issue at the same time.

Reply:
@alxiong Since `MerkleTreeGadget` is a trait, that becomes a little problematic. You either need to use `dyn MerkleTreeGadget` here instead, or do something like:
```
impl<T> DigestAlgorithmGadget for T
where
    T: MerkleTreeGadget
```

which I'm not even sure is possible, since we need to pass in the parameter `F` somehow like `impl<T, F>` (because we want to restrict `T: MerkleTreeGadget<RescueMerkleTree<F>>`), but the trait `DigestAlgorithmGadget` doesn't itself restrict `F`, and we'd run into errors.
By all means please drop a few lines/PR if you have an alternative of how to implement `DigestAlgorithmGadget` only for `MerkleTreeGadget` and not entire `PlonkCircuit` in this design.

But before you do that, please first checkout https://github.com/EspressoSystems/jellyfish/pull/167/commits/371e87470b2261ee1ea459d9f7e87a5d634d0397, which has an alternative design, and actually avoids the issue you mention altogether. (note that the concrete location of the structs/impls there is up for discussion, I might move them later to where `RescueNativeGadget` is defined).

Reply:
I kind of like your alternative solution, particularly we don't have the `dyn` issue, also compliant with my intuition that `DigestGadget` should be associated/part of/used under the context of `MerkleTreeGadget`.